### DISCUSSION TITLE
Calculate total number of packages to install before installing

### DISCUSSION THREAD

**Discussion by vjpr** [pnpm/pnpm#3174] [Ideas]
I wonder if there is a way to download the entire dependency tree first so that we can provide an accurate progress bar.

I am assuming it was an early idea but was found to be too slow...

It would allow us to have a install time remaining estimate too.

**Comment by zkochan** [comment]
This is impossible to do with pnpm. pnpm is verry aggressive and will be even more aggressive once we finish this PR: https://github.com/pnpm/supi/pull/8.

Different parts of the dependency tree can be on completely different installation stages. pnpm doesn't have shared installation stages like Yarn. Yarn downloads all the meta first, then all the tarballs, then creates the node_modules folders. pnpm does it all as early as possible for each package. Because we can do it, we can start writing the node_modules very early, everything in our symlinked node_modules is very deterministic

**Comment by vjpr** [comment]
So I guess pulling the metadata upfront would introduce a delay. 

Maybe it would be cool to have a separate package that could take a `shrinkwrap.yaml` and `package.json` and return the count of the packages. Then these could be chained together if necessary. But out of scope for this repo...

**Comment by zkochan** [comment]
I think it is fine to discuss all pnpm-related things in this repo. Yes, probably we can do some precalculations if there is an up-to-date `shrinkwrap.yaml` file
